##### Copyright 2020 The TensorFlow Authors.

In [1]:
#@title Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# https://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

# Multi-task recommenders

<table class="tfo-notebook-buttons" align="left">
  <td>
    <a target="_blank" href="https://www.tensorflow.org/recommenders/examples/multitask"><img src="https://www.tensorflow.org/images/tf_logo_32px.png" />View on TensorFlow.org</a>
  </td>
  <td>
    <a target="_blank" href="https://colab.research.google.com/github/tensorflow/recommenders/blob/main/docs/examples/multitask.ipynb"><img src="https://www.tensorflow.org/images/colab_logo_32px.png" />Run in Google Colab</a>
  </td>
  <td>
    <a target="_blank" href="https://github.com/tensorflow/recommenders/blob/main/docs/examples/multitask.ipynb"><img src="https://www.tensorflow.org/images/GitHub-Mark-32px.png" />View source on GitHub</a>
  </td>
  <td>
    <a href="https://storage.googleapis.com/tensorflow_docs/recommenders/docs/examples/multitask.ipynb"><img src="https://www.tensorflow.org/images/download_logo_32px.png" />Download notebook</a>
  </td>
</table>

In the [basic retrieval tutorial](basic_retrieval) we built a retrieval system using movie watches as positive interaction signals.

In many applications, however, there are multiple rich sources of feedback to draw upon. For example, an e-commerce site may record user visits to product pages (abundant, but relatively low signal), image clicks, adding to cart, and, finally, purchases. It may even record post-purchase signals such as reviews and returns.

Integrating all these different forms of feedback is critical to building systems that users love to use, and that do not optimize for any one metric at the expense of overall performance.

In addition, building a joint model for multiple tasks may produce better results than building a number of task-specific models. This is especially true where some data is abundant (for example, clicks), and some data is sparse (purchases, returns, manual reviews). In those scenarios, a joint model may be able to use representations learned from the abundant task to improve its predictions on the sparse task via a phenomenon known as [transfer learning](https://en.wikipedia.org/wiki/Transfer_learning). For example, [this paper](https://openreview.net/pdf?id=SJxPVcSonN) shows that a model predicting explicit user ratings from sparse user surveys can be substantially improved by adding an auxiliary task that uses abundant click log data.

In this tutorial, we are going to build a multi-objective recommender for Movielens, using both implicit (movie watches) and explicit signals (ratings).

## Imports


Let's first get our imports out of the way.


In [2]:
!pip install -q tensorflow-recommenders
!pip install -q --upgrade tensorflow-datasets

In [3]:
import os
import pprint
import tempfile

from typing import Dict, Text

import numpy as np
import tensorflow as tf
import tensorflow_datasets as tfds

In [4]:
import tensorflow_recommenders as tfrs

## Preparing the dataset

We're going to use the Movielens 100K dataset.

In [5]:
ratings = tfds.load('movielens/100k-ratings', split="train")
movies = tfds.load('movielens/100k-movies', split="train")

# Select the basic features.
ratings = ratings.map(lambda x: {
    "movie_title": x["movie_title"],
    "user_id": x["user_id"],
    "user_rating": x["user_rating"],
})
movies = movies.map(lambda x: x["movie_title"])

And repeat our preparations for building vocabularies and splitting the data into a train and a test set:

In [6]:
# Randomly shuffle data and split between train and test.
tf.random.set_seed(42)
shuffled = ratings.shuffle(100_000, seed=42, reshuffle_each_iteration=False)

train = shuffled.take(80_000)
test = shuffled.skip(80_000).take(20_000)

movie_titles = movies.batch(1_000)
user_ids = ratings.batch(1_000_000).map(lambda x: x["user_id"])

unique_movie_titles = np.unique(np.concatenate(list(movie_titles)))
unique_user_ids = np.unique(np.concatenate(list(user_ids)))

## A multi-task model

There are two critical parts to multi-task recommenders:

1. They optimize for two or more objectives, and so have two or more losses.
2. They share variables between the tasks, allowing for transfer learning.

In this tutorial, we will define our models as before, but instead of having  a single task, we will have two tasks: one that predicts ratings, and one that predicts movie watches.

The user and movie models are as before:

```python
user_model = tf.keras.Sequential([
  tf.keras.layers.StringLookup(
      vocabulary=unique_user_ids, mask_token=None),
  # We add 1 to account for the unknown token.
  tf.keras.layers.Embedding(len(unique_user_ids) + 1, embedding_dimension)
])

movie_model = tf.keras.Sequential([
  tf.keras.layers.StringLookup(
      vocabulary=unique_movie_titles, mask_token=None),
  tf.keras.layers.Embedding(len(unique_movie_titles) + 1, embedding_dimension)
])
```

However, now we will have two tasks. The first is the rating task:

```python
tfrs.tasks.Ranking(
    loss=tf.keras.losses.MeanSquaredError(),
    metrics=[tf.keras.metrics.RootMeanSquaredError()],
)
```

Its goal is to predict the ratings as accurately as possible.

The second is the retrieval task:

```python
tfrs.tasks.Retrieval(
    metrics=tfrs.metrics.FactorizedTopK(
        candidates=movies.batch(128)
    )
)
```

As before, this task's goal is to predict which movies the user will or will not watch.

### Putting it together

We put it all together in a model class.

The new component here is that - since we have two tasks and two losses - we need to decide on how important each loss is. We can do this by giving each of the losses a weight, and treating these weights as hyperparameters. If we assign a large loss weight to the rating task, our model is going to focus on predicting ratings (but still use some information from the retrieval task); if we assign a large loss weight to the retrieval task, it will focus on retrieval instead.

In [7]:
class MovielensModel(tfrs.models.Model):

  def __init__(self, rating_weight: float, retrieval_weight: float) -> None:
    # We take the loss weights in the constructor: this allows us to instantiate
    # several model objects with different loss weights.

    super().__init__()

    embedding_dimension = 32

    # User and movie models.
    self.movie_model: tf.keras.layers.Layer = tf.keras.Sequential([
      tf.keras.layers.StringLookup(
        vocabulary=unique_movie_titles, mask_token=None),
      tf.keras.layers.Embedding(len(unique_movie_titles) + 1, embedding_dimension)
    ])
    self.user_model: tf.keras.layers.Layer = tf.keras.Sequential([
      tf.keras.layers.StringLookup(
        vocabulary=unique_user_ids, mask_token=None),
      tf.keras.layers.Embedding(len(unique_user_ids) + 1, embedding_dimension)
    ])

    # A small model to take in user and movie embeddings and predict ratings.
    # We can make this as complicated as we want as long as we output a scalar
    # as our prediction.
    self.rating_model = tf.keras.Sequential([
        tf.keras.layers.Dense(256, activation="relu"),
        tf.keras.layers.Dense(128, activation="relu"),
        tf.keras.layers.Dense(1),
    ])

    # The tasks.
    self.rating_task: tf.keras.layers.Layer = tfrs.tasks.Ranking(
        loss=tf.keras.losses.MeanSquaredError(),
        metrics=[tf.keras.metrics.RootMeanSquaredError()],
    )
    self.retrieval_task: tf.keras.layers.Layer = tfrs.tasks.Retrieval(
        metrics=tfrs.metrics.FactorizedTopK(
            candidates=movies.batch(128).map(self.movie_model)
        )
    )

    # The loss weights.
    self.rating_weight = rating_weight
    self.retrieval_weight = retrieval_weight

  def call(self, features: Dict[Text, tf.Tensor]) -> tf.Tensor:
    # We pick out the user features and pass them into the user model.
    user_embeddings = self.user_model(features["user_id"])
    # And pick out the movie features and pass them into the movie model.
    movie_embeddings = self.movie_model(features["movie_title"])
    
    return (
        user_embeddings,
        movie_embeddings,
        # We apply the multi-layered rating model to a concatentation of
        # user and movie embeddings.
        self.rating_model(
            tf.concat([user_embeddings, movie_embeddings], axis=1)
        ),
    )

  def compute_loss(self, features: Dict[Text, tf.Tensor], training=False) -> tf.Tensor:

    ratings = features.pop("user_rating")

    user_embeddings, movie_embeddings, rating_predictions = self(features)

    # We compute the loss for each task.
    rating_loss = self.rating_task(
        labels=ratings,
        predictions=rating_predictions,
    )
    retrieval_loss = self.retrieval_task(user_embeddings, movie_embeddings)

    # And combine them using the loss weights.
    return (self.rating_weight * rating_loss
            + self.retrieval_weight * retrieval_loss)

### Rating-specialized model

Depending on the weights we assign, the model will encode a different balance of the tasks. Let's start with a model that only considers ratings.

In [8]:
model = MovielensModel(rating_weight=1.0, retrieval_weight=0.0)
model.compile(optimizer=tf.keras.optimizers.Adagrad(0.1))

In [9]:
cached_train = train.shuffle(100_000).batch(8192).cache()
cached_test = test.batch(4096).cache()

In [10]:
model.fit(cached_train, epochs=3)
metrics = model.evaluate(cached_test, return_dict=True)

print(f"Retrieval top-100 accuracy: {metrics['factorized_top_k/top_100_categorical_accuracy']:.3f}.")
print(f"Ranking RMSE: {metrics['root_mean_squared_error']:.3f}.")

Epoch 1/3


 1/10 [==>...........................] - ETA: 35s - root_mean_squared_error: 3.6922 - factorized_top_k/top_1_categorical_accuracy: 4.8828e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0028 - factorized_top_k/top_10_categorical_accuracy: 0.0061 - factorized_top_k/top_50_categorical_accuracy: 0.0293 - factorized_top_k/top_100_categorical_accuracy: 0.0576 - loss: 13.6320 - regularization_loss: 0.0000e+00 - total_loss: 13.6320

 2/10 [=====>........................] - ETA: 2s - root_mean_squared_error: 3.2709 - factorized_top_k/top_1_categorical_accuracy: 4.2725e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0028 - factorized_top_k/top_10_categorical_accuracy: 0.0059 - factorized_top_k/top_50_categorical_accuracy: 0.0295 - factorized_top_k/top_100_categorical_accuracy: 0.0587 - loss: 10.6989 - regularization_loss: 0.0000e+00 - total_loss: 10.6989 

 3/10 [========>.....................] - ETA: 2s - root_mean_squared_error: 2.8313 - factorized_top_k/top_1_categorical_accuracy: 4.0690e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0027 - factorized_top_k/top_10_categorical_accuracy: 0.0060 - factorized_top_k/top_50_categorical_accuracy: 0.0304 - factorized_top_k/top_100_categorical_accuracy: 0.0594 - loss: 8.0165 - regularization_loss: 0.0000e+00 - total_loss: 8.0165  

 4/10 [===========>..................] - ETA: 2s - root_mean_squared_error: 2.8111 - factorized_top_k/top_1_categorical_accuracy: 3.3569e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0025 - factorized_top_k/top_10_categorical_accuracy: 0.0058 - factorized_top_k/top_50_categorical_accuracy: 0.0298 - factorized_top_k/top_100_categorical_accuracy: 0.0595 - loss: 7.9023 - regularization_loss: 0.0000e+00 - total_loss: 7.9023

 5/10 [==============>...............] - ETA: 1s - root_mean_squared_error: 2.6035 - factorized_top_k/top_1_categorical_accuracy: 3.4180e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0025 - factorized_top_k/top_10_categorical_accuracy: 0.0057 - factorized_top_k/top_50_categorical_accuracy: 0.0291 - factorized_top_k/top_100_categorical_accuracy: 0.0580 - loss: 6.7782 - regularization_loss: 0.0000e+00 - total_loss: 6.7782

 6/10 [=================>............] - ETA: 1s - root_mean_squared_error: 2.4424 - factorized_top_k/top_1_categorical_accuracy: 3.6621e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0026 - factorized_top_k/top_10_categorical_accuracy: 0.0057 - factorized_top_k/top_50_categorical_accuracy: 0.0297 - factorized_top_k/top_100_categorical_accuracy: 0.0593 - loss: 5.9655 - regularization_loss: 0.0000e+00 - total_loss: 5.9655

 7/10 [====================>.........] - ETA: 1s - root_mean_squared_error: 2.3456 - factorized_top_k/top_1_categorical_accuracy: 3.3133e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0025 - factorized_top_k/top_10_categorical_accuracy: 0.0056 - factorized_top_k/top_50_categorical_accuracy: 0.0294 - factorized_top_k/top_100_categorical_accuracy: 0.0589 - loss: 5.5017 - regularization_loss: 0.0000e+00 - total_loss: 5.5017

 8/10 [=======================>......] - ETA: 0s - root_mean_squared_error: 2.2365 - factorized_top_k/top_1_categorical_accuracy: 3.0518e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0025 - factorized_top_k/top_10_categorical_accuracy: 0.0055 - factorized_top_k/top_50_categorical_accuracy: 0.0293 - factorized_top_k/top_100_categorical_accuracy: 0.0585 - loss: 5.0019 - regularization_loss: 0.0000e+00 - total_loss: 5.0019

 9/10 [==========================>...] - ETA: 0s - root_mean_squared_error: 2.1507 - factorized_top_k/top_1_categorical_accuracy: 2.8483e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0024 - factorized_top_k/top_10_categorical_accuracy: 0.0054 - factorized_top_k/top_50_categorical_accuracy: 0.0293 - factorized_top_k/top_100_categorical_accuracy: 0.0587 - loss: 4.6257 - regularization_loss: 0.0000e+00 - total_loss: 4.6257

10/10 [==============================] - ETA: 0s - root_mean_squared_error: 2.0903 - factorized_top_k/top_1_categorical_accuracy: 2.7500e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0024 - factorized_top_k/top_10_categorical_accuracy: 0.0054 - factorized_top_k/top_50_categorical_accuracy: 0.0294 - factorized_top_k/top_100_categorical_accuracy: 0.0589 - loss: 4.2989 - regularization_loss: 0.0000e+00 - total_loss: 4.2989

10/10 [==============================] - 7s 331ms/step - root_mean_squared_error: 2.0903 - factorized_top_k/top_1_categorical_accuracy: 2.7500e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0024 - factorized_top_k/top_10_categorical_accuracy: 0.0054 - factorized_top_k/top_50_categorical_accuracy: 0.0294 - factorized_top_k/top_100_categorical_accuracy: 0.0589 - loss: 4.0315 - regularization_loss: 0.0000e+00 - total_loss: 4.0315


Epoch 2/3


 1/10 [==>...........................] - ETA: 3s - root_mean_squared_error: 1.2016 - factorized_top_k/top_1_categorical_accuracy: 1.2207e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0024 - factorized_top_k/top_10_categorical_accuracy: 0.0063 - factorized_top_k/top_50_categorical_accuracy: 0.0287 - factorized_top_k/top_100_categorical_accuracy: 0.0573 - loss: 1.4439 - regularization_loss: 0.0000e+00 - total_loss: 1.4439

 2/10 [=====>........................] - ETA: 2s - root_mean_squared_error: 1.1857 - factorized_top_k/top_1_categorical_accuracy: 1.2207e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0027 - factorized_top_k/top_10_categorical_accuracy: 0.0057 - factorized_top_k/top_50_categorical_accuracy: 0.0293 - factorized_top_k/top_100_categorical_accuracy: 0.0587 - loss: 1.4059 - regularization_loss: 0.0000e+00 - total_loss: 1.4059

 3/10 [========>.....................] - ETA: 2s - root_mean_squared_error: 1.1732 - factorized_top_k/top_1_categorical_accuracy: 1.6276e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0028 - factorized_top_k/top_10_categorical_accuracy: 0.0058 - factorized_top_k/top_50_categorical_accuracy: 0.0305 - factorized_top_k/top_100_categorical_accuracy: 0.0596 - loss: 1.3763 - regularization_loss: 0.0000e+00 - total_loss: 1.3763

 4/10 [===========>..................] - ETA: 1s - root_mean_squared_error: 1.1660 - factorized_top_k/top_1_categorical_accuracy: 1.5259e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0026 - factorized_top_k/top_10_categorical_accuracy: 0.0057 - factorized_top_k/top_50_categorical_accuracy: 0.0302 - factorized_top_k/top_100_categorical_accuracy: 0.0599 - loss: 1.3596 - regularization_loss: 0.0000e+00 - total_loss: 1.3596

 5/10 [==============>...............] - ETA: 1s - root_mean_squared_error: 1.1633 - factorized_top_k/top_1_categorical_accuracy: 1.9531e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0026 - factorized_top_k/top_10_categorical_accuracy: 0.0057 - factorized_top_k/top_50_categorical_accuracy: 0.0293 - factorized_top_k/top_100_categorical_accuracy: 0.0586 - loss: 1.3533 - regularization_loss: 0.0000e+00 - total_loss: 1.3533

 6/10 [=================>............] - ETA: 1s - root_mean_squared_error: 1.1605 - factorized_top_k/top_1_categorical_accuracy: 2.4414e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0026 - factorized_top_k/top_10_categorical_accuracy: 0.0058 - factorized_top_k/top_50_categorical_accuracy: 0.0301 - factorized_top_k/top_100_categorical_accuracy: 0.0597 - loss: 1.3467 - regularization_loss: 0.0000e+00 - total_loss: 1.3467

 7/10 [====================>.........] - ETA: 0s - root_mean_squared_error: 1.1581 - factorized_top_k/top_1_categorical_accuracy: 2.2670e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0025 - factorized_top_k/top_10_categorical_accuracy: 0.0056 - factorized_top_k/top_50_categorical_accuracy: 0.0298 - factorized_top_k/top_100_categorical_accuracy: 0.0593 - loss: 1.3411 - regularization_loss: 0.0000e+00 - total_loss: 1.3411

 8/10 [=======================>......] - ETA: 0s - root_mean_squared_error: 1.1574 - factorized_top_k/top_1_categorical_accuracy: 2.1362e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0025 - factorized_top_k/top_10_categorical_accuracy: 0.0055 - factorized_top_k/top_50_categorical_accuracy: 0.0296 - factorized_top_k/top_100_categorical_accuracy: 0.0589 - loss: 1.3396 - regularization_loss: 0.0000e+00 - total_loss: 1.3396

 9/10 [==========================>...] - ETA: 0s - root_mean_squared_error: 1.1567 - factorized_top_k/top_1_categorical_accuracy: 2.0345e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0024 - factorized_top_k/top_10_categorical_accuracy: 0.0054 - factorized_top_k/top_50_categorical_accuracy: 0.0296 - factorized_top_k/top_100_categorical_accuracy: 0.0590 - loss: 1.3379 - regularization_loss: 0.0000e+00 - total_loss: 1.3379

10/10 [==============================] - ETA: 0s - root_mean_squared_error: 1.1531 - factorized_top_k/top_1_categorical_accuracy: 1.8750e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0024 - factorized_top_k/top_10_categorical_accuracy: 0.0054 - factorized_top_k/top_50_categorical_accuracy: 0.0297 - factorized_top_k/top_100_categorical_accuracy: 0.0591 - loss: 1.3275 - regularization_loss: 0.0000e+00 - total_loss: 1.3275

10/10 [==============================] - 3s 321ms/step - root_mean_squared_error: 1.1531 - factorized_top_k/top_1_categorical_accuracy: 1.8750e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0024 - factorized_top_k/top_10_categorical_accuracy: 0.0054 - factorized_top_k/top_50_categorical_accuracy: 0.0297 - factorized_top_k/top_100_categorical_accuracy: 0.0591 - loss: 1.3189 - regularization_loss: 0.0000e+00 - total_loss: 1.3189


Epoch 3/3


 1/10 [==>...........................] - ETA: 2s - root_mean_squared_error: 1.1343 - factorized_top_k/top_1_categorical_accuracy: 1.2207e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0026 - factorized_top_k/top_10_categorical_accuracy: 0.0060 - factorized_top_k/top_50_categorical_accuracy: 0.0288 - factorized_top_k/top_100_categorical_accuracy: 0.0585 - loss: 1.2867 - regularization_loss: 0.0000e+00 - total_loss: 1.2867

 2/10 [=====>........................] - ETA: 2s - root_mean_squared_error: 1.1330 - factorized_top_k/top_1_categorical_accuracy: 1.2207e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0026 - factorized_top_k/top_10_categorical_accuracy: 0.0057 - factorized_top_k/top_50_categorical_accuracy: 0.0294 - factorized_top_k/top_100_categorical_accuracy: 0.0593 - loss: 1.2838 - regularization_loss: 0.0000e+00 - total_loss: 1.2838

 3/10 [========>.....................] - ETA: 2s - root_mean_squared_error: 1.1222 - factorized_top_k/top_1_categorical_accuracy: 1.6276e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0028 - factorized_top_k/top_10_categorical_accuracy: 0.0059 - factorized_top_k/top_50_categorical_accuracy: 0.0307 - factorized_top_k/top_100_categorical_accuracy: 0.0601 - loss: 1.2593 - regularization_loss: 0.0000e+00 - total_loss: 1.2593

 4/10 [===========>..................] - ETA: 1s - root_mean_squared_error: 1.1203 - factorized_top_k/top_1_categorical_accuracy: 1.5259e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0026 - factorized_top_k/top_10_categorical_accuracy: 0.0057 - factorized_top_k/top_50_categorical_accuracy: 0.0303 - factorized_top_k/top_100_categorical_accuracy: 0.0605 - loss: 1.2550 - regularization_loss: 0.0000e+00 - total_loss: 1.2550

 5/10 [==============>...............] - ETA: 1s - root_mean_squared_error: 1.1199 - factorized_top_k/top_1_categorical_accuracy: 1.7090e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0026 - factorized_top_k/top_10_categorical_accuracy: 0.0057 - factorized_top_k/top_50_categorical_accuracy: 0.0295 - factorized_top_k/top_100_categorical_accuracy: 0.0594 - loss: 1.2543 - regularization_loss: 0.0000e+00 - total_loss: 1.2543

 6/10 [=================>............] - ETA: 1s - root_mean_squared_error: 1.1206 - factorized_top_k/top_1_categorical_accuracy: 1.8311e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0026 - factorized_top_k/top_10_categorical_accuracy: 0.0058 - factorized_top_k/top_50_categorical_accuracy: 0.0303 - factorized_top_k/top_100_categorical_accuracy: 0.0602 - loss: 1.2557 - regularization_loss: 0.0000e+00 - total_loss: 1.2557

 7/10 [====================>.........] - ETA: 0s - root_mean_squared_error: 1.1202 - factorized_top_k/top_1_categorical_accuracy: 1.7439e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0026 - factorized_top_k/top_10_categorical_accuracy: 0.0057 - factorized_top_k/top_50_categorical_accuracy: 0.0302 - factorized_top_k/top_100_categorical_accuracy: 0.0599 - loss: 1.2549 - regularization_loss: 0.0000e+00 - total_loss: 1.2549

 8/10 [=======================>......] - ETA: 0s - root_mean_squared_error: 1.1214 - factorized_top_k/top_1_categorical_accuracy: 1.8311e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0026 - factorized_top_k/top_10_categorical_accuracy: 0.0056 - factorized_top_k/top_50_categorical_accuracy: 0.0301 - factorized_top_k/top_100_categorical_accuracy: 0.0595 - loss: 1.2575 - regularization_loss: 0.0000e+00 - total_loss: 1.2575

 9/10 [==========================>...] - ETA: 0s - root_mean_squared_error: 1.1219 - factorized_top_k/top_1_categorical_accuracy: 1.7632e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0025 - factorized_top_k/top_10_categorical_accuracy: 0.0055 - factorized_top_k/top_50_categorical_accuracy: 0.0299 - factorized_top_k/top_100_categorical_accuracy: 0.0596 - loss: 1.2586 - regularization_loss: 0.0000e+00 - total_loss: 1.2586

10/10 [==============================] - ETA: 0s - root_mean_squared_error: 1.1198 - factorized_top_k/top_1_categorical_accuracy: 1.6250e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0025 - factorized_top_k/top_10_categorical_accuracy: 0.0055 - factorized_top_k/top_50_categorical_accuracy: 0.0300 - factorized_top_k/top_100_categorical_accuracy: 0.0597 - loss: 1.2527 - regularization_loss: 0.0000e+00 - total_loss: 1.2527

10/10 [==============================] - 3s 316ms/step - root_mean_squared_error: 1.1198 - factorized_top_k/top_1_categorical_accuracy: 1.6250e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0025 - factorized_top_k/top_10_categorical_accuracy: 0.0055 - factorized_top_k/top_50_categorical_accuracy: 0.0300 - factorized_top_k/top_100_categorical_accuracy: 0.0597 - loss: 1.2479 - regularization_loss: 0.0000e+00 - total_loss: 1.2479


1/5 [=====>........................] - ETA: 8s - root_mean_squared_error: 1.1173 - factorized_top_k/top_1_categorical_accuracy: 4.8828e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0027 - factorized_top_k/top_10_categorical_accuracy: 0.0042 - factorized_top_k/top_50_categorical_accuracy: 0.0266 - factorized_top_k/top_100_categorical_accuracy: 0.0603 - loss: 1.2483 - regularization_loss: 0.0000e+00 - total_loss: 1.2483

2/5 [===========>..................] - ETA: 0s - root_mean_squared_error: 1.1210 - factorized_top_k/top_1_categorical_accuracy: 4.8828e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0026 - factorized_top_k/top_10_categorical_accuracy: 0.0044 - factorized_top_k/top_50_categorical_accuracy: 0.0264 - factorized_top_k/top_100_categorical_accuracy: 0.0574 - loss: 1.2565 - regularization_loss: 0.0000e+00 - total_loss: 1.2565

3/5 [=================>............] - ETA: 0s - root_mean_squared_error: 1.1200 - factorized_top_k/top_1_categorical_accuracy: 5.6966e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0027 - factorized_top_k/top_10_categorical_accuracy: 0.0046 - factorized_top_k/top_50_categorical_accuracy: 0.0278 - factorized_top_k/top_100_categorical_accuracy: 0.0579 - loss: 1.2543 - regularization_loss: 0.0000e+00 - total_loss: 1.2543

4/5 [=======================>......] - ETA: 0s - root_mean_squared_error: 1.1156 - factorized_top_k/top_1_categorical_accuracy: 4.8828e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0029 - factorized_top_k/top_10_categorical_accuracy: 0.0050 - factorized_top_k/top_50_categorical_accuracy: 0.0291 - factorized_top_k/top_100_categorical_accuracy: 0.0598 - loss: 1.2447 - regularization_loss: 0.0000e+00 - total_loss: 1.2447

5/5 [==============================] - ETA: 0s - root_mean_squared_error: 1.1130 - factorized_top_k/top_1_categorical_accuracy: 4.5000e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0028 - factorized_top_k/top_10_categorical_accuracy: 0.0052 - factorized_top_k/top_50_categorical_accuracy: 0.0295 - factorized_top_k/top_100_categorical_accuracy: 0.0597 - loss: 1.2380 - regularization_loss: 0.0000e+00 - total_loss: 1.2380

5/5 [==============================] - 3s 194ms/step - root_mean_squared_error: 1.1130 - factorized_top_k/top_1_categorical_accuracy: 4.5000e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0028 - factorized_top_k/top_10_categorical_accuracy: 0.0052 - factorized_top_k/top_50_categorical_accuracy: 0.0295 - factorized_top_k/top_100_categorical_accuracy: 0.0597 - loss: 1.2336 - regularization_loss: 0.0000e+00 - total_loss: 1.2336


Retrieval top-100 accuracy: 0.060.
Ranking RMSE: 1.113.


The model does OK on predicting ratings (with an RMSE of around 1.11), but performs poorly at predicting which movies will be watched or not: its accuracy at 100 is almost 4 times worse than a model trained solely to predict watches.

### Retrieval-specialized model

Let's now try a model that focuses on retrieval only.

In [11]:
model = MovielensModel(rating_weight=0.0, retrieval_weight=1.0)
model.compile(optimizer=tf.keras.optimizers.Adagrad(0.1))

In [12]:
model.fit(cached_train, epochs=3)
metrics = model.evaluate(cached_test, return_dict=True)

print(f"Retrieval top-100 accuracy: {metrics['factorized_top_k/top_100_categorical_accuracy']:.3f}.")
print(f"Ranking RMSE: {metrics['root_mean_squared_error']:.3f}.")

Epoch 1/3


 1/10 [==>...........................] - ETA: 9s - root_mean_squared_error: 3.6975 - factorized_top_k/top_1_categorical_accuracy: 0.0000e+00 - factorized_top_k/top_5_categorical_accuracy: 1.2207e-04 - factorized_top_k/top_10_categorical_accuracy: 6.1035e-04 - factorized_top_k/top_50_categorical_accuracy: 0.0112 - factorized_top_k/top_100_categorical_accuracy: 0.0320 - loss: 73817.5703 - regularization_loss: 0.0000e+00 - total_loss: 73817.5703

 2/10 [=====>........................] - ETA: 2s - root_mean_squared_error: 3.7038 - factorized_top_k/top_1_categorical_accuracy: 0.0000e+00 - factorized_top_k/top_5_categorical_accuracy: 3.6621e-04 - factorized_top_k/top_10_categorical_accuracy: 0.0012 - factorized_top_k/top_50_categorical_accuracy: 0.0156 - factorized_top_k/top_100_categorical_accuracy: 0.0401 - loss: 73818.4453 - regularization_loss: 0.0000e+00 - total_loss: 73818.4453    

 3/10 [========>.....................] - ETA: 2s - root_mean_squared_error: 3.7059 - factorized_top_k/top_1_categorical_accuracy: 0.0000e+00 - factorized_top_k/top_5_categorical_accuracy: 5.2897e-04 - factorized_top_k/top_10_categorical_accuracy: 0.0019 - factorized_top_k/top_50_categorical_accuracy: 0.0199 - factorized_top_k/top_100_categorical_accuracy: 0.0486 - loss: 73819.6042 - regularization_loss: 0.0000e+00 - total_loss: 73819.6042

 4/10 [===========>..................] - ETA: 1s - root_mean_squared_error: 3.7098 - factorized_top_k/top_1_categorical_accuracy: 3.0518e-05 - factorized_top_k/top_5_categorical_accuracy: 8.8501e-04 - factorized_top_k/top_10_categorical_accuracy: 0.0027 - factorized_top_k/top_50_categorical_accuracy: 0.0233 - factorized_top_k/top_100_categorical_accuracy: 0.0561 - loss: 73815.9863 - regularization_loss: 0.0000e+00 - total_loss: 73815.9863

 5/10 [==============>...............] - ETA: 1s - root_mean_squared_error: 3.7128 - factorized_top_k/top_1_categorical_accuracy: 2.4414e-05 - factorized_top_k/top_5_categorical_accuracy: 9.5215e-04 - factorized_top_k/top_10_categorical_accuracy: 0.0029 - factorized_top_k/top_50_categorical_accuracy: 0.0256 - factorized_top_k/top_100_categorical_accuracy: 0.0631 - loss: 73799.9406 - regularization_loss: 0.0000e+00 - total_loss: 73799.9406

 6/10 [=================>............] - ETA: 1s - root_mean_squared_error: 3.7169 - factorized_top_k/top_1_categorical_accuracy: 2.0345e-05 - factorized_top_k/top_5_categorical_accuracy: 9.1553e-04 - factorized_top_k/top_10_categorical_accuracy: 0.0028 - factorized_top_k/top_50_categorical_accuracy: 0.0281 - factorized_top_k/top_100_categorical_accuracy: 0.0726 - loss: 73756.4193 - regularization_loss: 0.0000e+00 - total_loss: 73756.4193

 7/10 [====================>.........] - ETA: 0s - root_mean_squared_error: 3.7211 - factorized_top_k/top_1_categorical_accuracy: 1.7439e-05 - factorized_top_k/top_5_categorical_accuracy: 8.8937e-04 - factorized_top_k/top_10_categorical_accuracy: 0.0028 - factorized_top_k/top_50_categorical_accuracy: 0.0315 - factorized_top_k/top_100_categorical_accuracy: 0.0819 - loss: 73678.2143 - regularization_loss: 0.0000e+00 - total_loss: 73678.2143

 8/10 [=======================>......] - ETA: 0s - root_mean_squared_error: 3.7208 - factorized_top_k/top_1_categorical_accuracy: 4.5776e-05 - factorized_top_k/top_5_categorical_accuracy: 0.0010 - factorized_top_k/top_10_categorical_accuracy: 0.0030 - factorized_top_k/top_50_categorical_accuracy: 0.0361 - factorized_top_k/top_100_categorical_accuracy: 0.0933 - loss: 73566.4346 - regularization_loss: 0.0000e+00 - total_loss: 73566.4346    

 9/10 [==========================>...] - ETA: 0s - root_mean_squared_error: 3.7217 - factorized_top_k/top_1_categorical_accuracy: 4.0690e-05 - factorized_top_k/top_5_categorical_accuracy: 0.0012 - factorized_top_k/top_10_categorical_accuracy: 0.0035 - factorized_top_k/top_50_categorical_accuracy: 0.0420 - factorized_top_k/top_100_categorical_accuracy: 0.1044 - loss: 73441.5234 - regularization_loss: 0.0000e+00 - total_loss: 73441.5234

10/10 [==============================] - ETA: 0s - root_mean_squared_error: 3.7238 - factorized_top_k/top_1_categorical_accuracy: 7.5000e-05 - factorized_top_k/top_5_categorical_accuracy: 0.0014 - factorized_top_k/top_10_categorical_accuracy: 0.0041 - factorized_top_k/top_50_categorical_accuracy: 0.0473 - factorized_top_k/top_100_categorical_accuracy: 0.1135 - loss: 71448.6020 - regularization_loss: 0.0000e+00 - total_loss: 71448.6020

10/10 [==============================] - 4s 313ms/step - root_mean_squared_error: 3.7238 - factorized_top_k/top_1_categorical_accuracy: 7.5000e-05 - factorized_top_k/top_5_categorical_accuracy: 0.0014 - factorized_top_k/top_10_categorical_accuracy: 0.0041 - factorized_top_k/top_50_categorical_accuracy: 0.0473 - factorized_top_k/top_100_categorical_accuracy: 0.1135 - loss: 69818.0298 - regularization_loss: 0.0000e+00 - total_loss: 69818.0298


Epoch 2/3


 1/10 [==>...........................] - ETA: 2s - root_mean_squared_error: 3.7272 - factorized_top_k/top_1_categorical_accuracy: 2.4414e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0046 - factorized_top_k/top_10_categorical_accuracy: 0.0134 - factorized_top_k/top_50_categorical_accuracy: 0.1144 - factorized_top_k/top_100_categorical_accuracy: 0.2272 - loss: 71781.0859 - regularization_loss: 0.0000e+00 - total_loss: 71781.0859

 2/10 [=====>........................] - ETA: 2s - root_mean_squared_error: 3.7329 - factorized_top_k/top_1_categorical_accuracy: 6.1035e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0063 - factorized_top_k/top_10_categorical_accuracy: 0.0159 - factorized_top_k/top_50_categorical_accuracy: 0.1166 - factorized_top_k/top_100_categorical_accuracy: 0.2353 - loss: 71630.4102 - regularization_loss: 0.0000e+00 - total_loss: 71630.4102

 3/10 [========>.....................] - ETA: 2s - root_mean_squared_error: 3.7344 - factorized_top_k/top_1_categorical_accuracy: 5.6966e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0074 - factorized_top_k/top_10_categorical_accuracy: 0.0187 - factorized_top_k/top_50_categorical_accuracy: 0.1212 - factorized_top_k/top_100_categorical_accuracy: 0.2426 - loss: 71485.6172 - regularization_loss: 0.0000e+00 - total_loss: 71485.6172

 4/10 [===========>..................] - ETA: 2s - root_mean_squared_error: 3.7380 - factorized_top_k/top_1_categorical_accuracy: 8.5449e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0083 - factorized_top_k/top_10_categorical_accuracy: 0.0201 - factorized_top_k/top_50_categorical_accuracy: 0.1273 - factorized_top_k/top_100_categorical_accuracy: 0.2497 - loss: 71338.7910 - regularization_loss: 0.0000e+00 - total_loss: 71338.7910

 5/10 [==============>...............] - ETA: 1s - root_mean_squared_error: 3.7409 - factorized_top_k/top_1_categorical_accuracy: 8.5449e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0093 - factorized_top_k/top_10_categorical_accuracy: 0.0222 - factorized_top_k/top_50_categorical_accuracy: 0.1314 - factorized_top_k/top_100_categorical_accuracy: 0.2532 - loss: 71245.5328 - regularization_loss: 0.0000e+00 - total_loss: 71245.5328

 6/10 [=================>............] - ETA: 1s - root_mean_squared_error: 3.7447 - factorized_top_k/top_1_categorical_accuracy: 8.7484e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0099 - factorized_top_k/top_10_categorical_accuracy: 0.0237 - factorized_top_k/top_50_categorical_accuracy: 0.1345 - factorized_top_k/top_100_categorical_accuracy: 0.2568 - loss: 71141.9049 - regularization_loss: 0.0000e+00 - total_loss: 71141.9049

 7/10 [====================>.........] - ETA: 1s - root_mean_squared_error: 3.7484 - factorized_top_k/top_1_categorical_accuracy: 9.5912e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0104 - factorized_top_k/top_10_categorical_accuracy: 0.0247 - factorized_top_k/top_50_categorical_accuracy: 0.1373 - factorized_top_k/top_100_categorical_accuracy: 0.2599 - loss: 71057.7801 - regularization_loss: 0.0000e+00 - total_loss: 71057.7801

 8/10 [=======================>......] - ETA: 0s - root_mean_squared_error: 3.7475 - factorized_top_k/top_1_categorical_accuracy: 0.0010 - factorized_top_k/top_5_categorical_accuracy: 0.0110 - factorized_top_k/top_10_categorical_accuracy: 0.0258 - factorized_top_k/top_50_categorical_accuracy: 0.1394 - factorized_top_k/top_100_categorical_accuracy: 0.2623 - loss: 70959.7207 - regularization_loss: 0.0000e+00 - total_loss: 70959.7207    

 9/10 [==========================>...] - ETA: 0s - root_mean_squared_error: 3.7479 - factorized_top_k/top_1_categorical_accuracy: 0.0011 - factorized_top_k/top_5_categorical_accuracy: 0.0115 - factorized_top_k/top_10_categorical_accuracy: 0.0264 - factorized_top_k/top_50_categorical_accuracy: 0.1412 - factorized_top_k/top_100_categorical_accuracy: 0.2644 - loss: 70895.4792 - regularization_loss: 0.0000e+00 - total_loss: 70895.4792

10/10 [==============================] - ETA: 0s - root_mean_squared_error: 3.7495 - factorized_top_k/top_1_categorical_accuracy: 0.0011 - factorized_top_k/top_5_categorical_accuracy: 0.0116 - factorized_top_k/top_10_categorical_accuracy: 0.0268 - factorized_top_k/top_50_categorical_accuracy: 0.1425 - factorized_top_k/top_100_categorical_accuracy: 0.2658 - loss: 69013.2742 - regularization_loss: 0.0000e+00 - total_loss: 69013.2742

10/10 [==============================] - 3s 326ms/step - root_mean_squared_error: 3.7495 - factorized_top_k/top_1_categorical_accuracy: 0.0011 - factorized_top_k/top_5_categorical_accuracy: 0.0116 - factorized_top_k/top_10_categorical_accuracy: 0.0268 - factorized_top_k/top_50_categorical_accuracy: 0.1425 - factorized_top_k/top_100_categorical_accuracy: 0.2658 - loss: 67473.2884 - regularization_loss: 0.0000e+00 - total_loss: 67473.2884


Epoch 3/3


 1/10 [==>...........................] - ETA: 3s - root_mean_squared_error: 3.7464 - factorized_top_k/top_1_categorical_accuracy: 8.5449e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0153 - factorized_top_k/top_10_categorical_accuracy: 0.0350 - factorized_top_k/top_50_categorical_accuracy: 0.1638 - factorized_top_k/top_100_categorical_accuracy: 0.2897 - loss: 70042.7031 - regularization_loss: 0.0000e+00 - total_loss: 70042.7031

 2/10 [=====>........................] - ETA: 2s - root_mean_squared_error: 3.7518 - factorized_top_k/top_1_categorical_accuracy: 0.0013 - factorized_top_k/top_5_categorical_accuracy: 0.0156 - factorized_top_k/top_10_categorical_accuracy: 0.0351 - factorized_top_k/top_50_categorical_accuracy: 0.1644 - factorized_top_k/top_100_categorical_accuracy: 0.2910 - loss: 69991.5781 - regularization_loss: 0.0000e+00 - total_loss: 69991.5781    

 3/10 [========>.....................] - ETA: 2s - root_mean_squared_error: 3.7529 - factorized_top_k/top_1_categorical_accuracy: 0.0012 - factorized_top_k/top_5_categorical_accuracy: 0.0162 - factorized_top_k/top_10_categorical_accuracy: 0.0361 - factorized_top_k/top_50_categorical_accuracy: 0.1689 - factorized_top_k/top_100_categorical_accuracy: 0.2948 - loss: 69916.0781 - regularization_loss: 0.0000e+00 - total_loss: 69916.0781

 4/10 [===========>..................] - ETA: 1s - root_mean_squared_error: 3.7561 - factorized_top_k/top_1_categorical_accuracy: 0.0014 - factorized_top_k/top_5_categorical_accuracy: 0.0171 - factorized_top_k/top_10_categorical_accuracy: 0.0378 - factorized_top_k/top_50_categorical_accuracy: 0.1723 - factorized_top_k/top_100_categorical_accuracy: 0.2993 - loss: 69835.9551 - regularization_loss: 0.0000e+00 - total_loss: 69835.9551

 5/10 [==============>...............] - ETA: 1s - root_mean_squared_error: 3.7584 - factorized_top_k/top_1_categorical_accuracy: 0.0015 - factorized_top_k/top_5_categorical_accuracy: 0.0179 - factorized_top_k/top_10_categorical_accuracy: 0.0386 - factorized_top_k/top_50_categorical_accuracy: 0.1738 - factorized_top_k/top_100_categorical_accuracy: 0.3005 - loss: 69798.6484 - regularization_loss: 0.0000e+00 - total_loss: 69798.6484

 6/10 [=================>............] - ETA: 1s - root_mean_squared_error: 3.7618 - factorized_top_k/top_1_categorical_accuracy: 0.0015 - factorized_top_k/top_5_categorical_accuracy: 0.0180 - factorized_top_k/top_10_categorical_accuracy: 0.0390 - factorized_top_k/top_50_categorical_accuracy: 0.1756 - factorized_top_k/top_100_categorical_accuracy: 0.3029 - loss: 69744.9102 - regularization_loss: 0.0000e+00 - total_loss: 69744.9102

 7/10 [====================>.........] - ETA: 0s - root_mean_squared_error: 3.7648 - factorized_top_k/top_1_categorical_accuracy: 0.0015 - factorized_top_k/top_5_categorical_accuracy: 0.0182 - factorized_top_k/top_10_categorical_accuracy: 0.0387 - factorized_top_k/top_50_categorical_accuracy: 0.1761 - factorized_top_k/top_100_categorical_accuracy: 0.3031 - loss: 69717.8895 - regularization_loss: 0.0000e+00 - total_loss: 69717.8895

 8/10 [=======================>......] - ETA: 0s - root_mean_squared_error: 3.7635 - factorized_top_k/top_1_categorical_accuracy: 0.0014 - factorized_top_k/top_5_categorical_accuracy: 0.0181 - factorized_top_k/top_10_categorical_accuracy: 0.0387 - factorized_top_k/top_50_categorical_accuracy: 0.1764 - factorized_top_k/top_100_categorical_accuracy: 0.3045 - loss: 69664.0820 - regularization_loss: 0.0000e+00 - total_loss: 69664.0820

 9/10 [==========================>...] - ETA: 0s - root_mean_squared_error: 3.7635 - factorized_top_k/top_1_categorical_accuracy: 0.0014 - factorized_top_k/top_5_categorical_accuracy: 0.0181 - factorized_top_k/top_10_categorical_accuracy: 0.0388 - factorized_top_k/top_50_categorical_accuracy: 0.1767 - factorized_top_k/top_100_categorical_accuracy: 0.3048 - loss: 69644.1250 - regularization_loss: 0.0000e+00 - total_loss: 69644.1250

10/10 [==============================] - ETA: 0s - root_mean_squared_error: 3.7648 - factorized_top_k/top_1_categorical_accuracy: 0.0014 - factorized_top_k/top_5_categorical_accuracy: 0.0180 - factorized_top_k/top_10_categorical_accuracy: 0.0388 - factorized_top_k/top_50_categorical_accuracy: 0.1773 - factorized_top_k/top_100_categorical_accuracy: 0.3050 - loss: 67820.9461 - regularization_loss: 0.0000e+00 - total_loss: 67820.9461

10/10 [==============================] - 3s 314ms/step - root_mean_squared_error: 3.7648 - factorized_top_k/top_1_categorical_accuracy: 0.0014 - factorized_top_k/top_5_categorical_accuracy: 0.0180 - factorized_top_k/top_10_categorical_accuracy: 0.0388 - factorized_top_k/top_50_categorical_accuracy: 0.1773 - factorized_top_k/top_100_categorical_accuracy: 0.3050 - loss: 66329.2543 - regularization_loss: 0.0000e+00 - total_loss: 66329.2543


1/5 [=====>........................] - ETA: 2s - root_mean_squared_error: 3.7755 - factorized_top_k/top_1_categorical_accuracy: 9.7656e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0117 - factorized_top_k/top_10_categorical_accuracy: 0.0256 - factorized_top_k/top_50_categorical_accuracy: 0.1274 - factorized_top_k/top_100_categorical_accuracy: 0.2388 - loss: 32458.2539 - regularization_loss: 0.0000e+00 - total_loss: 32458.2539

2/5 [===========>..................] - ETA: 0s - root_mean_squared_error: 3.7762 - factorized_top_k/top_1_categorical_accuracy: 6.1035e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0098 - factorized_top_k/top_10_categorical_accuracy: 0.0215 - factorized_top_k/top_50_categorical_accuracy: 0.1243 - factorized_top_k/top_100_categorical_accuracy: 0.2361 - loss: 32508.5273 - regularization_loss: 0.0000e+00 - total_loss: 32508.5273

3/5 [=================>............] - ETA: 0s - root_mean_squared_error: 3.7744 - factorized_top_k/top_1_categorical_accuracy: 8.1380e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0092 - factorized_top_k/top_10_categorical_accuracy: 0.0216 - factorized_top_k/top_50_categorical_accuracy: 0.1252 - factorized_top_k/top_100_categorical_accuracy: 0.2346 - loss: 32516.4395 - regularization_loss: 0.0000e+00 - total_loss: 32516.4395

4/5 [=======================>......] - ETA: 0s - root_mean_squared_error: 3.7732 - factorized_top_k/top_1_categorical_accuracy: 0.0012 - factorized_top_k/top_5_categorical_accuracy: 0.0093 - factorized_top_k/top_10_categorical_accuracy: 0.0217 - factorized_top_k/top_50_categorical_accuracy: 0.1245 - factorized_top_k/top_100_categorical_accuracy: 0.2358 - loss: 32495.7471 - regularization_loss: 0.0000e+00 - total_loss: 32495.7471    

5/5 [==============================] - ETA: 0s - root_mean_squared_error: 3.7730 - factorized_top_k/top_1_categorical_accuracy: 0.0012 - factorized_top_k/top_5_categorical_accuracy: 0.0097 - factorized_top_k/top_10_categorical_accuracy: 0.0218 - factorized_top_k/top_50_categorical_accuracy: 0.1253 - factorized_top_k/top_100_categorical_accuracy: 0.2352 - loss: 31649.3406 - regularization_loss: 0.0000e+00 - total_loss: 31649.3406

5/5 [==============================] - 1s 193ms/step - root_mean_squared_error: 3.7730 - factorized_top_k/top_1_categorical_accuracy: 0.0012 - factorized_top_k/top_5_categorical_accuracy: 0.0097 - factorized_top_k/top_10_categorical_accuracy: 0.0218 - factorized_top_k/top_50_categorical_accuracy: 0.1253 - factorized_top_k/top_100_categorical_accuracy: 0.2352 - loss: 31085.0697 - regularization_loss: 0.0000e+00 - total_loss: 31085.0697


Retrieval top-100 accuracy: 0.235.
Ranking RMSE: 3.773.


We get the opposite result: a model that does well on retrieval, but poorly on predicting ratings.

### Joint model

Let's now train a model that assigns positive weights to both tasks.

In [13]:
model = MovielensModel(rating_weight=1.0, retrieval_weight=1.0)
model.compile(optimizer=tf.keras.optimizers.Adagrad(0.1))

In [14]:
model.fit(cached_train, epochs=3)
metrics = model.evaluate(cached_test, return_dict=True)

print(f"Retrieval top-100 accuracy: {metrics['factorized_top_k/top_100_categorical_accuracy']:.3f}.")
print(f"Ranking RMSE: {metrics['root_mean_squared_error']:.3f}.")

Epoch 1/3


 1/10 [==>...........................] - ETA: 9s - root_mean_squared_error: 3.6615 - factorized_top_k/top_1_categorical_accuracy: 0.0000e+00 - factorized_top_k/top_5_categorical_accuracy: 2.4414e-04 - factorized_top_k/top_10_categorical_accuracy: 4.8828e-04 - factorized_top_k/top_50_categorical_accuracy: 0.0116 - factorized_top_k/top_100_categorical_accuracy: 0.0353 - loss: 73830.4453 - regularization_loss: 0.0000e+00 - total_loss: 73830.4453

 2/10 [=====>........................] - ETA: 2s - root_mean_squared_error: 3.1084 - factorized_top_k/top_1_categorical_accuracy: 6.1035e-05 - factorized_top_k/top_5_categorical_accuracy: 6.7139e-04 - factorized_top_k/top_10_categorical_accuracy: 0.0018 - factorized_top_k/top_50_categorical_accuracy: 0.0173 - factorized_top_k/top_100_categorical_accuracy: 0.0431 - loss: 73829.0469 - regularization_loss: 0.0000e+00 - total_loss: 73829.0469    

 3/10 [========>.....................] - ETA: 2s - root_mean_squared_error: 3.0812 - factorized_top_k/top_1_categorical_accuracy: 4.0690e-05 - factorized_top_k/top_5_categorical_accuracy: 8.9518e-04 - factorized_top_k/top_10_categorical_accuracy: 0.0022 - factorized_top_k/top_50_categorical_accuracy: 0.0211 - factorized_top_k/top_100_categorical_accuracy: 0.0507 - loss: 73829.5234 - regularization_loss: 0.0000e+00 - total_loss: 73829.5234

 4/10 [===========>..................] - ETA: 1s - root_mean_squared_error: 3.1885 - factorized_top_k/top_1_categorical_accuracy: 6.1035e-05 - factorized_top_k/top_5_categorical_accuracy: 0.0010 - factorized_top_k/top_10_categorical_accuracy: 0.0025 - factorized_top_k/top_50_categorical_accuracy: 0.0248 - factorized_top_k/top_100_categorical_accuracy: 0.0591 - loss: 73825.9160 - regularization_loss: 0.0000e+00 - total_loss: 73825.9160    

 5/10 [==============>...............] - ETA: 1s - root_mean_squared_error: 3.1477 - factorized_top_k/top_1_categorical_accuracy: 4.8828e-05 - factorized_top_k/top_5_categorical_accuracy: 0.0010 - factorized_top_k/top_10_categorical_accuracy: 0.0025 - factorized_top_k/top_50_categorical_accuracy: 0.0274 - factorized_top_k/top_100_categorical_accuracy: 0.0657 - loss: 73809.7172 - regularization_loss: 0.0000e+00 - total_loss: 73809.7172

 6/10 [=================>............] - ETA: 1s - root_mean_squared_error: 3.0200 - factorized_top_k/top_1_categorical_accuracy: 4.0690e-05 - factorized_top_k/top_5_categorical_accuracy: 9.5622e-04 - factorized_top_k/top_10_categorical_accuracy: 0.0024 - factorized_top_k/top_50_categorical_accuracy: 0.0289 - factorized_top_k/top_100_categorical_accuracy: 0.0734 - loss: 73765.6224 - regularization_loss: 0.0000e+00 - total_loss: 73765.6224

 7/10 [====================>.........] - ETA: 0s - root_mean_squared_error: 2.8407 - factorized_top_k/top_1_categorical_accuracy: 3.4877e-05 - factorized_top_k/top_5_categorical_accuracy: 9.2425e-04 - factorized_top_k/top_10_categorical_accuracy: 0.0027 - factorized_top_k/top_50_categorical_accuracy: 0.0314 - factorized_top_k/top_100_categorical_accuracy: 0.0808 - loss: 73683.8627 - regularization_loss: 0.0000e+00 - total_loss: 73683.8627

 8/10 [=======================>......] - ETA: 0s - root_mean_squared_error: 2.6953 - factorized_top_k/top_1_categorical_accuracy: 3.0518e-05 - factorized_top_k/top_5_categorical_accuracy: 0.0010 - factorized_top_k/top_10_categorical_accuracy: 0.0031 - factorized_top_k/top_50_categorical_accuracy: 0.0346 - factorized_top_k/top_100_categorical_accuracy: 0.0902 - loss: 73568.7539 - regularization_loss: 0.0000e+00 - total_loss: 73568.7539    

 9/10 [==========================>...] - ETA: 0s - root_mean_squared_error: 2.5823 - factorized_top_k/top_1_categorical_accuracy: 2.7127e-05 - factorized_top_k/top_5_categorical_accuracy: 0.0012 - factorized_top_k/top_10_categorical_accuracy: 0.0036 - factorized_top_k/top_50_categorical_accuracy: 0.0400 - factorized_top_k/top_100_categorical_accuracy: 0.1011 - loss: 73439.2457 - regularization_loss: 0.0000e+00 - total_loss: 73439.2457

10/10 [==============================] - ETA: 0s - root_mean_squared_error: 2.5007 - factorized_top_k/top_1_categorical_accuracy: 3.7500e-05 - factorized_top_k/top_5_categorical_accuracy: 0.0014 - factorized_top_k/top_10_categorical_accuracy: 0.0043 - factorized_top_k/top_50_categorical_accuracy: 0.0450 - factorized_top_k/top_100_categorical_accuracy: 0.1102 - loss: 71444.1656 - regularization_loss: 0.0000e+00 - total_loss: 71444.1656

10/10 [==============================] - 4s 299ms/step - root_mean_squared_error: 2.5007 - factorized_top_k/top_1_categorical_accuracy: 3.7500e-05 - factorized_top_k/top_5_categorical_accuracy: 0.0014 - factorized_top_k/top_10_categorical_accuracy: 0.0043 - factorized_top_k/top_50_categorical_accuracy: 0.0450 - factorized_top_k/top_100_categorical_accuracy: 0.1102 - loss: 69811.8274 - regularization_loss: 0.0000e+00 - total_loss: 69811.8274


Epoch 2/3


 1/10 [==>...........................] - ETA: 2s - root_mean_squared_error: 1.2498 - factorized_top_k/top_1_categorical_accuracy: 1.2207e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0051 - factorized_top_k/top_10_categorical_accuracy: 0.0143 - factorized_top_k/top_50_categorical_accuracy: 0.1107 - factorized_top_k/top_100_categorical_accuracy: 0.2272 - loss: 71774.2812 - regularization_loss: 0.0000e+00 - total_loss: 71774.2812

 2/10 [=====>........................] - ETA: 2s - root_mean_squared_error: 1.2114 - factorized_top_k/top_1_categorical_accuracy: 7.3242e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0067 - factorized_top_k/top_10_categorical_accuracy: 0.0170 - factorized_top_k/top_50_categorical_accuracy: 0.1118 - factorized_top_k/top_100_categorical_accuracy: 0.2288 - loss: 71638.0859 - regularization_loss: 0.0000e+00 - total_loss: 71638.0859

 3/10 [========>.....................] - ETA: 2s - root_mean_squared_error: 1.2093 - factorized_top_k/top_1_categorical_accuracy: 5.6966e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0064 - factorized_top_k/top_10_categorical_accuracy: 0.0177 - factorized_top_k/top_50_categorical_accuracy: 0.1163 - factorized_top_k/top_100_categorical_accuracy: 0.2365 - loss: 71496.2917 - regularization_loss: 0.0000e+00 - total_loss: 71496.2917

 4/10 [===========>..................] - ETA: 1s - root_mean_squared_error: 1.2012 - factorized_top_k/top_1_categorical_accuracy: 9.1553e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0075 - factorized_top_k/top_10_categorical_accuracy: 0.0193 - factorized_top_k/top_50_categorical_accuracy: 0.1227 - factorized_top_k/top_100_categorical_accuracy: 0.2434 - loss: 71351.8535 - regularization_loss: 0.0000e+00 - total_loss: 71351.8535

 5/10 [==============>...............] - ETA: 1s - root_mean_squared_error: 1.2119 - factorized_top_k/top_1_categorical_accuracy: 8.7891e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0085 - factorized_top_k/top_10_categorical_accuracy: 0.0215 - factorized_top_k/top_50_categorical_accuracy: 0.1275 - factorized_top_k/top_100_categorical_accuracy: 0.2467 - loss: 71257.9047 - regularization_loss: 0.0000e+00 - total_loss: 71257.9047

 6/10 [=================>............] - ETA: 1s - root_mean_squared_error: 1.2056 - factorized_top_k/top_1_categorical_accuracy: 8.7484e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0091 - factorized_top_k/top_10_categorical_accuracy: 0.0228 - factorized_top_k/top_50_categorical_accuracy: 0.1310 - factorized_top_k/top_100_categorical_accuracy: 0.2509 - loss: 71155.8971 - regularization_loss: 0.0000e+00 - total_loss: 71155.8971

 7/10 [====================>.........] - ETA: 0s - root_mean_squared_error: 1.2128 - factorized_top_k/top_1_categorical_accuracy: 8.7193e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0097 - factorized_top_k/top_10_categorical_accuracy: 0.0238 - factorized_top_k/top_50_categorical_accuracy: 0.1336 - factorized_top_k/top_100_categorical_accuracy: 0.2534 - loss: 71071.9431 - regularization_loss: 0.0000e+00 - total_loss: 71071.9431

 8/10 [=======================>......] - ETA: 0s - root_mean_squared_error: 1.2100 - factorized_top_k/top_1_categorical_accuracy: 8.8501e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0103 - factorized_top_k/top_10_categorical_accuracy: 0.0248 - factorized_top_k/top_50_categorical_accuracy: 0.1356 - factorized_top_k/top_100_categorical_accuracy: 0.2564 - loss: 70974.7490 - regularization_loss: 0.0000e+00 - total_loss: 70974.7490

 9/10 [==========================>...] - ETA: 0s - root_mean_squared_error: 1.2166 - factorized_top_k/top_1_categorical_accuracy: 0.0010 - factorized_top_k/top_5_categorical_accuracy: 0.0107 - factorized_top_k/top_10_categorical_accuracy: 0.0251 - factorized_top_k/top_50_categorical_accuracy: 0.1368 - factorized_top_k/top_100_categorical_accuracy: 0.2586 - loss: 70909.9297 - regularization_loss: 0.0000e+00 - total_loss: 70909.9297    

10/10 [==============================] - ETA: 0s - root_mean_squared_error: 1.2097 - factorized_top_k/top_1_categorical_accuracy: 9.8750e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0110 - factorized_top_k/top_10_categorical_accuracy: 0.0255 - factorized_top_k/top_50_categorical_accuracy: 0.1385 - factorized_top_k/top_100_categorical_accuracy: 0.2605 - loss: 69024.1676 - regularization_loss: 0.0000e+00 - total_loss: 69024.1676

10/10 [==============================] - 3s 312ms/step - root_mean_squared_error: 1.2097 - factorized_top_k/top_1_categorical_accuracy: 9.8750e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0110 - factorized_top_k/top_10_categorical_accuracy: 0.0255 - factorized_top_k/top_50_categorical_accuracy: 0.1385 - factorized_top_k/top_100_categorical_accuracy: 0.2605 - loss: 67481.2713 - regularization_loss: 0.0000e+00 - total_loss: 67481.2713


Epoch 3/3


 1/10 [==>...........................] - ETA: 2s - root_mean_squared_error: 1.1662 - factorized_top_k/top_1_categorical_accuracy: 2.4414e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0131 - factorized_top_k/top_10_categorical_accuracy: 0.0342 - factorized_top_k/top_50_categorical_accuracy: 0.1644 - factorized_top_k/top_100_categorical_accuracy: 0.2877 - loss: 70021.1406 - regularization_loss: 0.0000e+00 - total_loss: 70021.1406

 2/10 [=====>........................] - ETA: 2s - root_mean_squared_error: 1.1380 - factorized_top_k/top_1_categorical_accuracy: 0.0012 - factorized_top_k/top_5_categorical_accuracy: 0.0146 - factorized_top_k/top_10_categorical_accuracy: 0.0341 - factorized_top_k/top_50_categorical_accuracy: 0.1611 - factorized_top_k/top_100_categorical_accuracy: 0.2877 - loss: 69972.3945 - regularization_loss: 0.0000e+00 - total_loss: 69972.3945    

 3/10 [========>.....................] - ETA: 2s - root_mean_squared_error: 1.1280 - factorized_top_k/top_1_categorical_accuracy: 0.0011 - factorized_top_k/top_5_categorical_accuracy: 0.0148 - factorized_top_k/top_10_categorical_accuracy: 0.0354 - factorized_top_k/top_50_categorical_accuracy: 0.1669 - factorized_top_k/top_100_categorical_accuracy: 0.2929 - loss: 69890.3646 - regularization_loss: 0.0000e+00 - total_loss: 69890.3646

 4/10 [===========>..................] - ETA: 1s - root_mean_squared_error: 1.1207 - factorized_top_k/top_1_categorical_accuracy: 0.0012 - factorized_top_k/top_5_categorical_accuracy: 0.0157 - factorized_top_k/top_10_categorical_accuracy: 0.0364 - factorized_top_k/top_50_categorical_accuracy: 0.1709 - factorized_top_k/top_100_categorical_accuracy: 0.2982 - loss: 69806.6367 - regularization_loss: 0.0000e+00 - total_loss: 69806.6367

 5/10 [==============>...............] - ETA: 1s - root_mean_squared_error: 1.1203 - factorized_top_k/top_1_categorical_accuracy: 0.0013 - factorized_top_k/top_5_categorical_accuracy: 0.0165 - factorized_top_k/top_10_categorical_accuracy: 0.0375 - factorized_top_k/top_50_categorical_accuracy: 0.1721 - factorized_top_k/top_100_categorical_accuracy: 0.2989 - loss: 69769.5063 - regularization_loss: 0.0000e+00 - total_loss: 69769.5063

 6/10 [=================>............] - ETA: 1s - root_mean_squared_error: 1.1157 - factorized_top_k/top_1_categorical_accuracy: 0.0012 - factorized_top_k/top_5_categorical_accuracy: 0.0168 - factorized_top_k/top_10_categorical_accuracy: 0.0379 - factorized_top_k/top_50_categorical_accuracy: 0.1737 - factorized_top_k/top_100_categorical_accuracy: 0.3017 - loss: 69715.3151 - regularization_loss: 0.0000e+00 - total_loss: 69715.3151

 7/10 [====================>.........] - ETA: 0s - root_mean_squared_error: 1.1170 - factorized_top_k/top_1_categorical_accuracy: 0.0012 - factorized_top_k/top_5_categorical_accuracy: 0.0172 - factorized_top_k/top_10_categorical_accuracy: 0.0378 - factorized_top_k/top_50_categorical_accuracy: 0.1740 - factorized_top_k/top_100_categorical_accuracy: 0.3023 - loss: 69686.5246 - regularization_loss: 0.0000e+00 - total_loss: 69686.5246

 8/10 [=======================>......] - ETA: 0s - root_mean_squared_error: 1.1169 - factorized_top_k/top_1_categorical_accuracy: 0.0012 - factorized_top_k/top_5_categorical_accuracy: 0.0174 - factorized_top_k/top_10_categorical_accuracy: 0.0381 - factorized_top_k/top_50_categorical_accuracy: 0.1748 - factorized_top_k/top_100_categorical_accuracy: 0.3032 - loss: 69635.3447 - regularization_loss: 0.0000e+00 - total_loss: 69635.3447

 9/10 [==========================>...] - ETA: 0s - root_mean_squared_error: 1.1223 - factorized_top_k/top_1_categorical_accuracy: 0.0012 - factorized_top_k/top_5_categorical_accuracy: 0.0175 - factorized_top_k/top_10_categorical_accuracy: 0.0380 - factorized_top_k/top_50_categorical_accuracy: 0.1752 - factorized_top_k/top_100_categorical_accuracy: 0.3035 - loss: 69617.1128 - regularization_loss: 0.0000e+00 - total_loss: 69617.1128

10/10 [==============================] - ETA: 0s - root_mean_squared_error: 1.1200 - factorized_top_k/top_1_categorical_accuracy: 0.0011 - factorized_top_k/top_5_categorical_accuracy: 0.0175 - factorized_top_k/top_10_categorical_accuracy: 0.0380 - factorized_top_k/top_50_categorical_accuracy: 0.1758 - factorized_top_k/top_100_categorical_accuracy: 0.3040 - loss: 67791.5633 - regularization_loss: 0.0000e+00 - total_loss: 67791.5633

10/10 [==============================] - 3s 305ms/step - root_mean_squared_error: 1.1200 - factorized_top_k/top_1_categorical_accuracy: 0.0011 - factorized_top_k/top_5_categorical_accuracy: 0.0175 - factorized_top_k/top_10_categorical_accuracy: 0.0380 - factorized_top_k/top_50_categorical_accuracy: 0.1758 - factorized_top_k/top_100_categorical_accuracy: 0.3040 - loss: 66297.9318 - regularization_loss: 0.0000e+00 - total_loss: 66297.9318


1/5 [=====>........................] - ETA: 2s - root_mean_squared_error: 1.1393 - factorized_top_k/top_1_categorical_accuracy: 7.3242e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0085 - factorized_top_k/top_10_categorical_accuracy: 0.0227 - factorized_top_k/top_50_categorical_accuracy: 0.1248 - factorized_top_k/top_100_categorical_accuracy: 0.2334 - loss: 32468.5859 - regularization_loss: 0.0000e+00 - total_loss: 32468.5859

2/5 [===========>..................] - ETA: 0s - root_mean_squared_error: 1.1398 - factorized_top_k/top_1_categorical_accuracy: 9.7656e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0085 - factorized_top_k/top_10_categorical_accuracy: 0.0216 - factorized_top_k/top_50_categorical_accuracy: 0.1219 - factorized_top_k/top_100_categorical_accuracy: 0.2311 - loss: 32500.5625 - regularization_loss: 0.0000e+00 - total_loss: 32500.5625

3/5 [=================>............] - ETA: 0s - root_mean_squared_error: 1.1364 - factorized_top_k/top_1_categorical_accuracy: 8.9518e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0083 - factorized_top_k/top_10_categorical_accuracy: 0.0223 - factorized_top_k/top_50_categorical_accuracy: 0.1237 - factorized_top_k/top_100_categorical_accuracy: 0.2320 - loss: 32499.3958 - regularization_loss: 0.0000e+00 - total_loss: 32499.3958

4/5 [=======================>......] - ETA: 0s - root_mean_squared_error: 1.1337 - factorized_top_k/top_1_categorical_accuracy: 9.7656e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0082 - factorized_top_k/top_10_categorical_accuracy: 0.0215 - factorized_top_k/top_50_categorical_accuracy: 0.1238 - factorized_top_k/top_100_categorical_accuracy: 0.2337 - loss: 32477.6870 - regularization_loss: 0.0000e+00 - total_loss: 32477.6870

5/5 [==============================] - ETA: 0s - root_mean_squared_error: 1.1312 - factorized_top_k/top_1_categorical_accuracy: 9.5000e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0083 - factorized_top_k/top_10_categorical_accuracy: 0.0220 - factorized_top_k/top_50_categorical_accuracy: 0.1248 - factorized_top_k/top_100_categorical_accuracy: 0.2347 - loss: 31628.7672 - regularization_loss: 0.0000e+00 - total_loss: 31628.7672

5/5 [==============================] - 1s 187ms/step - root_mean_squared_error: 1.1312 - factorized_top_k/top_1_categorical_accuracy: 9.5000e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0083 - factorized_top_k/top_10_categorical_accuracy: 0.0220 - factorized_top_k/top_50_categorical_accuracy: 0.1248 - factorized_top_k/top_100_categorical_accuracy: 0.2347 - loss: 31062.8206 - regularization_loss: 0.0000e+00 - total_loss: 31062.8206


Retrieval top-100 accuracy: 0.235.
Ranking RMSE: 1.131.


The result is a model that performs roughly as well on both tasks as each specialized model.

### Making prediction

We can use the trained multitask model to get trained user and movie embeddings, as well as the predicted rating:

In [15]:
trained_movie_embeddings, trained_user_embeddings, predicted_rating = model({
      "user_id": np.array(["42"]),
      "movie_title": np.array(["Dances with Wolves (1990)"])
  })
print("Predicted rating:")
print(predicted_rating)

Predicted rating:
tf.Tensor([[3.4021819]], shape=(1, 1), dtype=float32)


While the results here do not show a clear accuracy benefit from a joint model in this case, multi-task learning is in general an extremely useful tool. We can expect better results when we can transfer knowledge from a data-abundant task (such as clicks) to a closely related data-sparse task (such as purchases).